docs/design/proposals/direct-blk-device-assignment.md

```diff
@@ -0,0 +1,246 @@
+implementing CSI volume resize and volume stat collection APIs.
+
+This document particularly focuses on how to address these two gaps.
+
+## Assumptions
+The proposal assumes that a block device volume will only be used by one Pod on a node at a time, which we believe 
+is the more common pattern in the Kata container. It’s rare and unsafe to have the same block device attached to more than 
+one Kata Pods. In the context of Kubernetes, the PersistentVolumeClaim (PVC) needs to have accessMode as `ReadWriteOncePod`. 
+
+## End User Interface
+If a PV is intended to be consumed by a VM-based runtime, the user can specify this by adding an annotation called 
+`skip-hostmount` to the persistent volume claim (PVC). Example:
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  annotations:
+    skip-hostmount: "true"
+  name: amazing-kata-volume
+spec:
+  accessModes:
+  - ReadWriteOncePod
+  resources:
+    requests:
+    storage: 4Gi
+```
+Based on this annotation, the CSI node driver knows it will not mount a filesystem on the volume and should defer the 
+operations to the container runtime. 
+
+## Interface between components
+
+Here we describe the changes we need to make in different components in order for the volume information to be propagated 
+from CSI to the Kata container runtime. 
+
+### Custom CSI node driver changes
+In the `skip-hostmount` mode, CSI node driver defers the filesystem mount operation to the container runtime. 
+The implementation of a few node driver APIs needs to be updated.
+
+* **NodePublishVolume** -- It propagates the mount information to the container runtime for it to carry out filesystem mount operation.
+* **NodeGetVolumeStats** -- Since the container runtime mounts the filesystem, it needs to forward the request to the container runtime.
+* **NodeExpandVolume** -- The CSI cannot implement this API alone. It needs to propagate the new size to the container runtime so that the filesystem metadata can be refreshed accordingly.
+* **NodeStageVolume/NodeUnStageVolume** -- Invoke a kata-runtime handler to clean up the stale mount information.
+
+### Kata-runtime changes
+The host agent (CSI node driver, as an example) will call into kata-runtime. Kata-runtime behaves as the API for 
+sandbox-level management interactions. In order for the CSI node driver to communicate with the kata-runtime, 
+we propose augmenting kata-runtime to include a filesystem, `fs`, command as follows:
+
+```
+kata-runtime fs —help
+NAME:
+kata-runtime fs [subcommand] [options]— interact with mounted filesystem volume
+
+SUBCOMMANDS:
+ - add-blk-mount-info <mountinfo> -- add the mount info of a volume to the Kata runtime.
+ - reset-blk-mount-info <host-blk-device-name> –- clean up the mount info directory under kata runtime
+ - get-stats <host-blk-device-name> — show status and usage details for the mounted filesystem backed by the given block device, <host-blk-device-name>.
```

Comment:
Can this also specify a return json to map the fields associated with CSI get stats (Condition, Size, Inode details, etc)?

Reply:
AFAIK, it's just the [interface](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/5d782870e3ee2e56e36202362d7b65d54ce8f866/vendor/github.com/container-storage-interface/spec/lib/go/csi/csi.pb.go#L4115-L4126) specified by the CSI, right?

Reply:
ok so the assumption here is that [1] this will be identical to the CSI get stats fields and [2] If CSI stats evolves with other fields, the structs/output of this will change correspondingly as well, correct?

Reply:
@fengwang666 Could you please answer @ddebroy's question above? Thanks ;-)

Reply:
@ddebroy Sorry for missing this comment. Yes. The assumptions are correct. @c3d Thanks for reminding me!